File: src/day-02/rock-paper-scissors.clj
(ns aoc22.day-02.rock-paper-scissors
  (:use [clojure.set :only [map-invert]])
  (:require [clojure.string :as s]))

(def ex1-encrypt {\A :rock \B :paper \C :scissor \X :rock \Y :paper \Z :scissor})
(def ex2-encrypt {\X :lose \Y :tie \Z :win})
(def win {:paper :rock :rock :scissor :scissor :paper})
(def lose (map-invert win))
(def points {:rock 1 :paper 2 :scissor 3 :win 6 :lose 0 :tie 3})
(def inp (slurp "src/day-02/inp.txt"))

(defn game-result [you adv]
  (cond (= (-> you win) adv)  :win
        (= adv you) :tie
        :else :lose))

(defn get-result-points [[adv you]] (+ (points you) (points (game-result you adv))))

(defn ex1 [text]
  (->> (s/split-lines text)
       (reduce (fn [acc line] (->> (seq line)
                                   ((juxt #(ex1-encrypt (first %))
                                          #(ex1-encrypt (nth % 2))))
                                   get-result-points
                                   (+ acc))) 0)))

(defn my-play [[adv result]]
  (cond (= :tie result) [adv adv]
        (= :win result) [adv (lose adv)]
        (= :lose result) [adv (win adv)]))

(defn ex2 [text]
  (->> (s/split-lines text)
       (reduce (fn [acc line] (->> (seq line)
                                   ((juxt #(ex1-encrypt (first %))
                                          #(ex2-encrypt (nth % 2))))
                                   my-play
                                   get-result-points
                                   (+ acc))) 0)))

(ex1 inp) ;; 14531
(ex2 inp) ;; 11258

File: src/day-02/rock_paper_scissors.ex
defmodule RockPaperScissor do
  @inp File.read!("./src/day-02/inp.txt")

  @ex1_encrypt %{
    "A" => :rock,
    "B" => :paper,
    "C" => :scissor,
    "X" => :rock,
    "Y" => :paper,
    "Z" => :scissor
  }

  @ex2_encrypt %{"X" => :lose, "Y" => :tie, "Z" => :win}
  @win %{rock: :scissor, paper: :rock, scissor: :paper}
  @points %{rock: 1, paper: 2, scissor: 3, win: 6, lose: 0, tie: 3}

  def exp1 do
    @inp
    |> String.split("\n")
    |> Enum.reduce(0, fn line, acc ->
      [adv, you] = String.split(line, " ")
      acc + round_points(@ex1_encrypt[adv], @ex1_encrypt[you])
    end)
  end

  def exp2 do
    @inp
    |> String.split("\n")
    |> Enum.reduce(0, fn line, acc ->
      [adv, expected_result] = String.split(line, " ")
      adv = @ex1_encrypt[adv]
      you = my_play_to_get_result(adv, @ex2_encrypt[expected_result])
      acc + round_points(adv, you)
    end)
  end

  defp my_play_to_get_result(adv, result) do
    case result do
      :tie -> adv
      :win -> @win[@win[adv]]
      :lose -> @win[adv]
    end
  end

  defp round_points(adv, you) do
    @points[you] + @points[game_result(adv, you)]
  end

  defp game_result(adv, you) do
    cond do
      you == adv -> :tie
      @win[you] == adv -> :win
      :else -> :lose
    end
  end
end
